include/merlin/core_kernels/contains.cuh

```diff
@@ -0,0 +1,340 @@
+    uint32_t target_digest = sm_target_digests[key_idx_block];
+    uint32_t target_digests = __byte_perm(target_digest, target_digest, 0x0000);
+    sm_counts[key_idx_block] = 0;
+    __pipeline_wait_prior(2);
+    uint32_t probing_digests =
+        sm_probing_digests[same_buf(i)][groupID * DIGEST_SPAN + rank];
+    uint32_t find_result_ = __vcmpeq4(probing_digests, target_digests);
+    uint32_t find_result = 0;
+    if ((find_result_ & 0x01) != 0) find_result |= 0x01;
+    if ((find_result_ & 0x0100) != 0) find_result |= 0x02;
+    if ((find_result_ & 0x010000) != 0) find_result |= 0x04;
+    if ((find_result_ & 0x01000000) != 0) find_result |= 0x08;
+    int find_number = __popc(find_result);
+    int group_base = 0;
+    if (find_number > 0) {
+      group_base = atomicAdd(sm_counts + key_idx_block, find_number);
+    }
+    bool gt_reserve = (group_base + find_number) > RESERVE;
+    int gt_vote = g.ballot(gt_reserve);
+    K* key_ptr = sm_keys_ptr[key_idx_block];
+    if (gt_vote == 0) {
+      do {
+        int digest_idx = __ffs(find_result) - 1;
+        if (digest_idx >= 0) {
+          find_result &= (find_result - 1);
+          int key_pos = rank * 4 + digest_idx;
+          __pipeline_memcpy_async(
+              sm_possible_keys[same_buf(i)] + (groupID * RESERVE + group_base),
+              key_ptr + key_pos, sizeof(K));
+          group_base += 1;
+        } else {
+          break;
+        }
+      } while (true);
+    } else {
+      K target_key = sm_target_keys[key_idx_block];
+      sm_counts[key_idx_block] = 0;
+      int found_vote = 0;
+      bool found = false;
+      do {
+        int digest_idx = __ffs(find_result) - 1;
+        if (digest_idx >= 0) {
+          find_result &= (find_result - 1);
+          int key_pos = rank * 4 + digest_idx;
+          K possible_key = key_ptr[key_pos];
+          if (possible_key == target_key) {
+            found = true;
+            sm_counts[key_idx_block] = 1;
+            sm_possible_keys[same_buf(i)][groupID * RESERVE] = possible_key;
+          }
+        }
+        found_vote = g.ballot(found);
+        if (found_vote) {
+          break;
+        }
+        found_vote = digest_idx >= 0;
+      } while (g.any(found_vote));
+    }
+    __pipeline_commit();
+
+    /* Step3: check possible keys, and prefecth the value and score */
+    if (i > 0) {
+      key_idx_block -= 1;
+      K target_key = sm_target_keys[key_idx_block];
+      int possible_num = sm_counts[key_idx_block];
+      sm_founds[key_idx_block] = 0;
+      __pipeline_wait_prior(2);
+      bool found_flag = false;
+      if (rank < possible_num) {
+        K possible_key =
+            sm_possible_keys[diff_buf(i)][groupID * RESERVE + rank];
+        if (possible_key == target_key) {
+          found_flag = true;
+        }
+      }
+      int found_vote = g.ballot(found_flag);
+      if (found_vote) {
+        sm_founds[key_idx_block] = 1;
+      }
+    }
+
+    /* Step4: write `found` */
```

Comment:
I think the step4 can be move into step 3. As the communication through shared memory is unnecessary.

Reply:
Agree, I has merged it.